src/utils/focus-management.ts

```diff
@@ -0,0 +1,183 @@
+// Credit: https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-vue/src/utils/focus-management.ts
+import { dom } from '~/utils/dom'
+
+import type { Ref } from '@nuxtjs/composition-api'
+
+function getOwnerDocument<T extends Element | Ref<Element | null>>(
+  element: T | null | undefined
+) {
+  if (typeof window === 'undefined') return null
+  if (element instanceof Node) return element.ownerDocument
+  // eslint-disable-next-line no-prototype-builtins
+  if (element?.hasOwnProperty('value')) {
+    const domElement = dom(element)
+    if (domElement) return domElement.ownerDocument
+  }
+
+  return document
+}
+
+// Credit:
+//  - https://stackoverflow.com/a/30753870
+const focusableSelector = [
+  '[contentEditable=true]',
+  '[tabindex]',
+  'a[href]',
+  'area[href]',
+  'button:not([disabled])',
+  'iframe',
+  'input:not([disabled])',
+  'select:not([disabled])',
+  'textarea:not([disabled])',
+]
+  .map(
+    process.env.NODE_ENV === 'test'
+      ? // TODO: Remove this once JSDOM fixes the issue where an element that is
+        // "hidden" can be the document.activeElement, because this is not possible
+        // in real browsers.
+        (selector) =>
+          `${selector}:not([tabindex='-1']):not([style*='display: none'])`
+      : (selector) => `${selector}:not([tabindex='-1'])`
+  )
+  .join(',')
+
+export enum Focus {
+  /** Focus the first non-disabled element */
+  First = 1 << 0,
+
+  /** Focus the previous non-disabled element */
+  Previous = 1 << 1,
+
+  /** Focus the next non-disabled element */
+  Next = 1 << 2,
+
+  /** Focus the last non-disabled element */
+  Last = 1 << 3,
+
+  /** Wrap tab around */
+  WrapAround = 1 << 4,
+}
+
+export enum FocusResult {
+  Error,
+  Overflow,
+  Success,
+  Underflow,
+}
+
+enum Direction {
+  Previous = -1,
+  Next = 1,
+}
```

Comment:
Let's avoid using the `enum` feature from TypeScript. It violates TS's promise of only adding types to JS but also adding new language features that JS doesn't have natively (same with the TS decorator implementation).

In these cases, we can just as easily use an object for the first one and a union for the other two:

```suggestion
export const Focus = Object.freeze({
  /** Focus the first non-disabled element */
  First: 1 << 0,

  /** Focus the previous non-disabled element */
  Previous: 1 << 1,

  /** Focus the next non-disabled element */
  Next: 1 << 2,

  /** Focus the last non-disabled element */
  Last: 1 << 3,

  /** Wrap tab around */
  WrapAround: 1 << 4,
}

export type FocusResult =
  | 'Error'
  | 'Overflow'
  | 'Success'
  | 'Underflow'

// This could be an object too if we want to keep the nice names.
type Direction =
  | -1
  | 1
```